### ISSUE TITLE
Arrays cannot be indexed by non-integral type 'String'

### ISSUE THREAD

**Issue by dragokas** [alliedmodders/sourcepawn#471]
```
#pragma newdecls required
#pragma semicolon 1

#include <sourcemod>

public void OnPluginStart()
{
	int c[256];
	char dest[] = "abcdefg";
	
	c[ dest[0] ] = dest[0];
}
```

```
SourcePawn Compiler 1.11.0.6526
Copyright (c) 1997-2006 ITB CompuPhase
Copyright (c) 2004-2018 AlliedModders LLC

C:\Users\Alex\Desktop\sp_parser.sp(11) : error 077: arrays cannot be indexed by non-integral type 'String'

1 Error.
```

Something similar was here: https://github.com/alliedmodders/sourcepawn/issues/401

**Comment by Accelerator74** [comment]
error 077: arrays cannot be indexed by non-integral type 'CBasePlayer'

`/**
 * List of operation types for clients arrays.
 **/
enum struct ClientData
{
	int Client_Respawn;
	int Client_RespawnTimes;
}

/**
 * Arrays to store the clients' data.
 **/
ClientData gClientData[MAXPLAYERS+1];

methodmap CBasePlayer
{


	//*********************************************************************
	//*            				CONSTRUCTOR           	  				  *
	//*********************************************************************

	//! Initialize methodmap class
	public CBasePlayer(int x)
	{
		return view_as<CBasePlayer>(x);
	}

	//! Transform index to normal integer
	property int Index 
	{ 
		public get()                
		{ 
			return view_as<int>(this); 
		} 
	}

	//! Respawn team variable
	property int m_bRespawn
	{
		public get() 
		{ 
			return gClientData[this].Client_Respawn; 
		}
		
		public set(int x) 
		{ 
			gClientData[this].Client_Respawn = x; 
		}
	}

	//! Respawn count variable
	property int m_nRespawnTimes
	{
		public get() 
		{ 
			return gClientData[this].Client_RespawnTimes; 
		}
		
		public set(int x) 
		{ 
			gClientData[this].Client_RespawnTimes = x;
		}
	}`

**Comment by dragokas** [comment]
C++ allows this. I see no reason why that should throw an error by design.

In such opinion:
```
int d = dest[0];
c[ d ] = dest[0];
```
should also be forbidden.

**Comment by dvander** [comment]
This isn't C++, nor is that a language that we're trying to emulate. Implicit char-to-int is forbidden in most languages because it leaks implementation details about encoding.

**Comment by dvander** [comment]
I'm still in wait-and-see mode on this. It'd be nice to keep it an error.

**Comment by Fyren** [comment]
> Implicit char-to-int is forbidden in most languages because it leaks implementation details about encoding.

You can still do `c['a'] = 1` and it doesn't even warn.  I think, at least, there's inconsistency there.

**Comment by dvander** [comment]
This has been fixed on 1.11-dev.
